### PR TITLE
RFC 198: Move the container image from Docker Hub to GitHub Registry

### PR DESCRIPTION
**Pull request by OrKoN**
This RFC proposes moving the container image from Docker Hub to GitHub Registry.

Issue https://github.com/web-platform-tests/wpt/issues/28903

[Preview](https://github.com/OrKoN/rfcs/blob/orkon/docker/rfcs/docker_github_registry.md)

### CODE REVIEW

File: rfcs/docker_github_registry.md
Comment (foolip) [inline rfcs/docker_github_registry.md:25]: Fixing the "chane" typo by rewording:

```suggestion
- We can take the opportunity to change the versioning scheme to use integer numbers.
```

File: rfcs/docker_github_registry.md
Comment (foolip) [inline rfcs/docker_github_registry.md:35]: I would guess the biggest risk is that Taskcluster pulling the image gets errors more often than with Docker Hub, or perhaps that some caching works differently so that all jobs take longer to start.

The reliability is hard to judge ahead of time, but the time it takes can be measured once at least to ensure no big difference.

**Review by foolip — APPROVED** [review]

File: rfcs/docker_github_registry.md
Comment (Ms2ger) [inline rfcs/docker_github_registry.md:19]: I don't know what this image is used for, but that's a lot of people who we haven't (and probably don't want to) vet for more than "is willing to review tests"

File: rfcs/docker_github_registry.md
Comment (OrKoN) [reply rfcs/docker_github_registry.md:35]: I think I can try to get the numbers from the test runs. I will add the information and update the RFC.

File: rfcs/docker_github_registry.md
Comment (OrKoN) [reply rfcs/docker_github_registry.md:19]: Screenshot 2024-05-29 at 10 19 34

Members with the Write access can make changes to the Dockerfile so it seems appropriate to also allow publishing the changes. Alternatively, on the package settings page there is an option to not inherit the permissions or disallow access from the workflows so the admin can change that and only allow selected members if that is desired.

Comment (jgraham) [reply rfcs/docker_github_registry.md:19]: We could use a different repo for the images if we wanted different permissions. 

We can also use `CODEOWNERS` to ensure that any changes to the docker image get admin review before being merged which would prevent updating the CI to use a new version of the image, but wouldn't prevent  publishing images in the first place.

We can also change the permissions on specific packages (notably the `wpt` package) to not inherit from the wpt base permissions. I think it would still (and maybe always was?) possible for people to publish new packages which would by default inherit permissions from the repo.

**Review by jcscottiii — COMMENTED** [review]
Hi @OrKoN 

We discussed this today in the infra [meeting](https://github.com/web-platform-tests/wpt-notes/blob/master/minutes/2024-06-04.md).

Overall this RFC looks good. Can you add a Permissions subsection to the Details section? In particular:

- Docker image pushing:
  - Disable default inheritance from the repository and provide access to:
    - The [admin](https://github.com/orgs/web-platform-tests/teams/admins) team
    - The [core](https://github.com/orgs/web-platform-tests/teams/wpt-core-team) team
    - the github workflows ([docs](https://docs.github.com/en/packages/learn-github-packages/about-permissions-for-github-packages#maintaining-access-to-packages-in-github-actions-workflows))
  - By default, a lot of individuals would have access to publish this package through the inherited permissions. This will limit the blast radius in case a certain account is compromised and pushes a package.
 - Use CODEOWNERS to require reviews from the admins and core teams for Dockerfile changes.

File: rfcs/docker_github_registry.md
Comment (OrKoN) [reply rfcs/docker_github_registry.md:35]: I have took two random TC jobs using the old and the new image and suprisingly the download time is ~48 seconds for both (I didn't see any indication of caching in the logs):

dockerhub registry: https://community-tc.services.mozilla.com/tasks/ZNnHobGNTiS4qBiv0l7Z-A/runs/0/logs/live/public/logs/live.log

github registry: https://community-tc.services.mozilla.com/tasks/F3k9fZ6TSdC6Cd9kIwnX0A/runs/1/logs/public/logs/live.log

**Comment by OrKoN** [comment]
@jcscottiii thanks for the feedback, I have included this into RFC. Note that someone with an admin permissions should configure the desired permissions (I have no access) and I can only add CODEOWNERS to my PR.

**Comment by OrKoN** [comment]
Screenshot 2024-06-04 at 20 03 58

The admins team does not seem to be suitable for CODEOWNERS. Perhaps it does not have write access?

File: rfcs/docker_github_registry.md
Comment (jgraham) [inline rfcs/docker_github_registry.md:44]: Can you be specific about which changes will require review? I think it should be changes to the Dockerfile and changes to the `.taskcluster.yml` file.

**Review by jgraham — APPROVED** [review]
Approved, but note the comment.

File: rfcs/docker_github_registry.md
Comment (OrKoN) [reply rfcs/docker_github_registry.md]: Updated

**Comment by foolip** [comment]
> Screenshot 2024-06-04 at 20 03 58
> The admins team does not seem to be suitable for CODEOWNERS. Perhaps it does not have write access?

Yeah, everyone in the admin team is made an org owner, so we never give the team itself write access to anything. I've given the team write access to the wpt repo now, can you try again?

**Comment by OrKoN** [comment]
@foolip I have updated my PR, and it works now. The admins can be specified in CODEOWNERS.

**Comment by OrKoN** [comment]
Shall we land this RFC? (I have no permission to merge, so if anyone could hit the merge button it would be great)

**Comment by jgraham** [comment]
I think technically we need to wait another day.

**Comment by OrKoN** [comment]
@foolip @jgraham @jcscottiii it looks like a bit more than 2 weeks already passed and there are not unresolved concerns. Therefore, if someone could merge the PR, it would be great.

**Pull request merged**
